[dtensor] refactor op dispatch and fix is_same_size/equal

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112931
* __->__ #112927

torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

wanchaol committed:
[dtensor] fix is_same_size/equal behavior

torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

[ghstack-poisoned]

wconstab commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    for arg_strategy in select_strategy.strategies:
+        arg_spec = arg_strategy.output_spec
+        if is_tensor_partial(arg_spec):
+            # if the arg_spec have partial, reshard to replicate
```
expand the comment to say 'why' partial ==> replicate?

wconstab commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    ]
+)
+def equal_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    # equal_strategy deals with ops that comparing two tensor, we need to make sure
```
is the idea here that before you can perform an equality test on 2 ops, you need 
 - (a) they have the same logical shape/size- otherwise can skip comparing data
 - (b) if a dim is sharded for one arg, but replicated for another arg, its nots straghtforward to memcmp their content locally
 - (c) so therefore we canonicalize the args by making them all replicated, so they are easy to compare locally


if so, then i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

wanchaol replied:
Yep this is a great summary, I'll put these into comments too!

> (a) they have the same logical shape/size- otherwise can skip comparing data
Yep this is a good point we can do such a optimization similar to is_same_size! but only compare the data when logical shape are the same

> i wonder if its more optimal to consider for (b) if both input args are sharded, but already have same sharding, could you skip the replicate step and do the comparison on a local-shard basis?

Yes this is what we are doing currently, currently I'm performing this check inside [sharding propagation](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/sharding_prop.py#L200), if all inputs are sharded in the same way, we don't need to do resharding to turn them as replicated. This strategy mainly try to follow one argument's sharding with the max number of shards, and see if every arg have the same sharding layout, if so, directly compare, otherwise, redistribute to the argument's sharding first.

wconstab reviewed (commented):
some questions mostly for my learning. maybe worth having someone else stamp but overall i think it looks right

wanchaol commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -257,6 +257,11 @@ def return_type_tuple_tensors(self) -> bool:
 
+    def return_type_tensor(self) -> bool:
+        return_types = self.op._schema.returns
+        # all dispatch ops only return Tensor or Tuple[Tensor], so this check if enough
```
Oh yes I should clarify in the comments: op only return tensor or tuple/list of tensors for tensor like return types, some op do return int/bool, therefore I need to check if the return type is a tensor type or not.

wanchaol commented on torch/distributed/_tensor/ops/tensor_ops.py:
```diff
@@ -59,6 +57,45 @@ def default_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+@register_op_strategy(
+    [
+        aten.equal.default,
+        aten.is_same_size.default,
```
yep is_same_size would never hit. I mainly leave it here because if logically these two share the same logic if we don't do any dispatching optimizations in this PR, and for completeness I was mostly thinking if we plan to explore "single graph expansion like dist compiler" in the future again, having all ops registered to one table might be better, but I can remove it if you feel we want to keep things consistent ;)

Because for `is_same_size` we don't need to run propagation as we can just directly compare the DTensor shapes, the dispatching logic is more like a optimization.

wconstab replied:
ok that convinced me, this is fine

XilunWu replied:
Do you think it's a good idea to keep all lists of ops (e.g. `skip_prop_list`, `random_op_list`) in one file for better readability?

wconstab commented on test/distributed/_tensor/test_tensor_ops.py:
```diff
@@ -253,6 +253,17 @@ def test_equal(self):
+            global_input, device_mesh, replica_spec, run_check=False
+        )
+
+        self.assertTrue(dist_tensor_1.equal(dist_tensor_3))
```
should we add tests for cases where values actually are different, and especially, for the corner case where the difference only occurs on a remote shard and we notice it by doing the replication?

wanchaol replied:
added.

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -125,6 +125,12 @@ def _operator_dispatch(
     sharding_propagator: ShardingPropagator,
 ) -> Tuple[object, OpSchema, OutputSharding]:
+    # operators that does not need to go through sharding propagation
+    if op_call == aten.is_same_size.default:
```
maybe not in this PR. But down the road, IMHO, we want to avoid this type check(for single or certain ops) as much as possible?

wanchaol replied:
I refactored the dispatch code a bit to not have the op type checks, now all things are handled by the `custom_op_handlers`

fduwjj reviewed (approved):
Unblock and please address Will's comment.

wanchaol committed:
Update on "[dtensor] fix is_same_size/equal behavior"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM. We can address Will's comment in a follow up PR.

wanchaol committed:
Update on "[dtensor] refactor op dispatch and fix is_same_size/equal"


torch.equal/is_same_size currently skips sharding prop and directly do
local tensor compute, this is wrong. for these two ops:

- torch.equal: should not skip sharding prop, need to have two DTensor
have the SAME sharding before compare local shard values
- torch.is_same_size: need to completely skip both sharding prop and
local compute

This PR refactors the existing op_dispatch to make it a class instance
so that we can do custom op handling, then fixes both torch.equal and
torch.is_same_size

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc

[ghstack-poisoned]

wanchaol commented:
> LGTM. We can address Will's comment in a follow up PR.

I think I already addressed them so it's good :)

wanchaol commented:
@pytorchbot merge
